gobblin-utility/src/main/java/gobblin/util/ProxiedFileSystemCache.java

```diff
@@ -0,0 +1,93 @@
+/*
+* Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+* this file except in compliance with the License. You may obtain a copy of the
+* License at  http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software distributed
+* under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+* CONDITIONS OF ANY KIND, either express or implied.
+*/
+
+package gobblin.util;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.security.token.Token;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+
+/**
+ * A cache for storing a mapping between Hadoop users and user {@link FileSystem} objects.
+ *
+ * <p>
+ *  This classes uses Guava's {@link Cache} for storing the user to {@link FileSystem} mapping, and creates the
+ *  {@link FileSystem}s using the {@link ProxiedFileSystemUtils} class.
+ * </p>
+ *
+ * @see {@link Cache}, {@link ProxiedFileSystemUtils}
+ */
+public class ProxiedFileSystemCache {
+
+  private static final int DEFAULT_MAX_CACHE_SIZE = 1000;
+
+  private static final Cache<String, FileSystem> USER_NAME_TO_FILESYSTEM_CACHE = CacheBuilder.newBuilder()
+      .maximumSize(DEFAULT_MAX_CACHE_SIZE).build();
+
+  /**
+   * Cached version of {@link ProxiedFileSystemUtils#getProxiedFileSystemUsingKeytab(String, String, Path, URI, Configuration)}.
+   */
+  public static FileSystem getProxiedFileSystemUsingKeytab(String userNameToProxyAs, String superUserName,
```

Comment:
Guava cache has a pattern for adding elements using a Callable (see https://code.google.com/p/guava-libraries/wiki/CachesExplained under the heading "From a Callable"). It might be cleaner for this function.